ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/AddBlock.hs

```diff
@@ -0,0 +1,193 @@
+
+import           Control.Monad.Class.MonadAsync
+import           Control.Monad.Class.MonadSTM
+import           Control.Monad.IOSim
+
+import           Control.Tracer
+
+import qualified Ouroboros.Network.AnchoredFragment as AF
+import           Ouroboros.Network.Chain (Chain)
+
+import           Ouroboros.Consensus.Ledger.Abstract
+import           Ouroboros.Consensus.Ledger.Extended
+import           Ouroboros.Consensus.Protocol.Abstract
+import           Ouroboros.Consensus.Util (chunks)
+import           Ouroboros.Consensus.Util.Condense (condense)
+import qualified Ouroboros.Consensus.Util.ThreadRegistry as ThreadRegistry
+
+import           Ouroboros.Storage.ChainDB (TraceAddBlockEvent (..), addBlock,
+                     closeDB, openDB, toChain)
+import qualified Ouroboros.Storage.ChainDB as ChainDB
+import qualified Ouroboros.Storage.ChainDB.Model as Model
+
+import           Test.Ouroboros.Storage.ChainDB.StateMachine (constrName,
+                     mkArgs)
+import           Test.Ouroboros.Storage.ChainDB.TestBlock
+
+tests :: TestTree
+tests = testGroup "AddBlock"
+    [ testProperty "addBlock multiple threads" prop_addBlock_multiple_threads
+    ]
+
+
+-- | Add a bunch of blocks from multiple threads and check whether the
+-- resulting chain is correct.
+--
+-- Why not use the parallel state machine tests for this? If one thread adds a
+-- block that fits onto the current block and the other thread adds a
+-- different block that also fits onto the current block, then the outcome
+-- depends on scheduling: either outcome is fine, but they will be different.
+-- quickcheck-state-machine expects the outcome to be deterministic.
+--
+-- Hence this test: don't check whether the resulting chain is equal to the
+-- model chain, but check whether they are equally preferable (candidates). In
+-- fact, it is more complicated than that. Say @k = 2@ and we first add the
+-- following blocks in the given order: @A@, @B@, @C@, @A'@, @B'@, @C'@, @D'@,
+-- then the current chain will end with @C@, even though the fork ending with
+-- @D'@ would be longer. The reason is for this is that we would have to roll
+-- back more @k@ blocks to switch to this fork. Now imagine that the last four
+-- blocks in the list above get added before the first three, then the current
+-- chain would have ended with @D'@. So the order in which blocks are added
+-- really matters.
+--
+-- We take the following approach to check whether the resulting chain is
+-- correct:
+--
+-- * We take the list of blocks to add
+-- * For each permutation of this list:
+--   - We add all the blocks in the permuted list to the model
+--   - We check whether the current chain of the model is equally preferable
+--     to the resulting chain. If so, the test passes.
+-- * If none of the permutations result in an equally preferable chain, the
+--   test fails.
+--
+-- TODO test with multiple protocols
+-- TODO test with different thread schedulings for the simulator
```

Comment:
Find a trace point (or multiple) in `addBlock` and add an arbitrary `threadDelay` there to get multiple interleavings.

Reply:
See #762